(DOCS-9611) Workflow connection groups (#26612)
DataDog/documentation: content/en/service_management/workflows/connections.md

To connect to an arbitrary service, use the HTTP connection type. For authentication options and setup instructions, see [HTTP action][10].

## Connection groups

You can create groups of connections so that your workflows can authenticate into the correct account or accounts based on the given inputs. Connections can be grouped together only if they share the same integration (for example, you cannot group GCP and AWS connections within the same group). 

> You define the members of a connection group using a connection's _Identifier Tags_. For example, you can create a connection group consisting of AWS accounts that have the `env` tag.

Reviewer: let's add `account_id: 123456789` as another example for AWS

Reviewer: Before your example below, let's add: Each connection in the group must have a set of unique tags, so that a workflow can dynamically select the correct correction at runtime.

Author: I actually replaced the existing example with account_id instead.

Each connection in the group must have a different set of tag values for the given Identifier Tags. For example:
- `connectionA {env:staging}` and `connectionB {env:prod}` can be grouped together.
- `connectionA {env:staging}` and `connectionC {env:staging}` can't be grouped, because the group would contain duplicate tag values.

> You can only add connections you have **Resolve Access** for, and you can only use connections within a group if you have Resolve Access for those connections. If you use a group with a connection to which you don't have Resolve Access _and_ when your workflow runs it points to that connection, the workflow fails with a `403 Forbidden` error. To fix this issue, you can:

Author: I need more information about what Resolve Access is and how you get it.

Reviewer: https://docs-staging.datadoghq.com/neko/DOCS-9611-wf-connection-groups/service_management/workflows/access/#restrict-access-on-a-specific-connection

> You can only add connections you have **Resolve Access** for, and you can only use connections within a group if you have Resolve Access for those connections. If you use a group with a connection to which you don't have Resolve Access _and_ when your workflow runs it points to that connection, the workflow fails with a `403 Forbidden` error. To fix this issue, you can:

Reviewer: "You can only add connections you have **Resolve Access** for, and you can only use connections within a group if you have Resolve Access for those connections."

Is there a way to simplify this sentence?

1. Enter a group name, then enter a set of up to three **Identifier Tags** that the connections you want to include in your group all have.
1. Under **Confirm Group**, use the checkboxes to select the specific members of your group.
1. Click **Next, Confirm Access**, then choose your desired access level for the group.
1. Click **Create**.

### Use a connection group

1. In your workflow, select an action that requires a connection.
1. In the **Connection** field, in the drop-down, select the desired connection group under **Groups**.
1. Fill in the desired values for the connection group **Identifiers**. For example, if your connection group is defined using the `env` Identifier Tag, and you have two environments, `prod` and `staging`, you could use either of those values (or an expression that evaluates to one of those values).
1. Fill in any other required step values, then click **Save** to save your workflow.

### Update a connection group

If you have edit access to a connection group, you can update the following attributes:
- Group name
- Identifier Tags (these can never be empty, but they can be completely replaced)
- Connections (a group can be empty)

> <div class="alert alert-warning"><strong>Note</strong>: After you create a connection group, you can't update the integration associated with the group.</div>

Reviewer: I'm not sure someone will ever want to do this, so i'm not sure it's worth calling out.

### Delete a connection group

To delete a connection group:

1. Hover over the group you want to delete and click the **delete (trash can)** icon.

> 1. Click **Delete**.

Reviewer: Add warning that this will impact any workflows currently using this group?

To connect to an arbitrary service, use the HTTP connection type. For authentication options and setup instructions, see [HTTP action][10].

> ## Connection groups

Reviewer 2: Thanks for incorporating [these changes](https://github.com/DataDog/documentation/pull/26672/files)! 
I see that the **Connection Tags** section is still left out. Is it planned to add it in a separate PR?
Connection tags are a part of this new feature and I believe it could be practical to describe them (especially their limitations) for global understanding of the group Identifier Tags, wdyt?

Author: I have added the section now, thanks!

You can create groups of connections so that your workflows can authenticate into the correct account or accounts based on the given inputs. Connections can be grouped together only if they share the same integration (for example, you cannot group GCP and AWS connections within the same group). 

You define the members of a connection group using a connection's _Identifier Tags_. For example, you can create a connection group consisting of AWS accounts that have the `account_id` tag.

> Each connection in the group must have a set of unique identifier tags so that a workflow can dynamically select the correct correction at runtime. For example:

Reviewer 2: nit
```suggestion
Each connection in the group must have a set of unique identifier tags so that a workflow can dynamically select the correct connection at runtime. For example:
```

To create a connection group:

1. Navigate to the [connections list][3].
1. On the left, click **Groups**.
1. Click **+ New Group**, then select an integration.
1. Enter a group name, then enter a set of up to three **Identifier Tags** that the connections you want to include in your group all have.
1. Under **Confirm Group**, use the checkboxes to select the specific members of your group.
1. Click **Next, Confirm Access**, then choose your desired access level for the group.
1. Click **Create**.

### Use a connection group

To use a connection group:

1. In your workflow, select an action that requires a connection.
1. In the **Connection** field, in the drop-down, select the desired connection group under **Groups**.
1. Fill in the desired values for the connection group **Identifiers**. For example, if your connection group is defined using the `env` Identifier Tag, and you have two environments, `prod` and `staging`, you could use either of those values (or an expression that evaluates to one of those values).
1. Fill in any other required step values, then click **Save** to save your workflow.

> **Note**: You can only use connections within a group if you have [Resolver permission][12] for those connections. If a workflow tries to use a connection you don't have Resolver permission for, the workflow fails with a `403 Forbidden` error. To fix this issue, you can:

Reviewer 2: this is nice! 👍

Author: Thanks! :)

- Configure your workflow so that it can't point to a connection you don't have Resolver permission for.
- Remove the connection you don't have Resolver permission for from the connection group.<div class="alert alert-danger"><strong>Note</strong>: If you are using a connection group for multiple workflows, removing a connection that another workflow relies on causes that workflow to fail.</div>

### Update a connection group

If you have edit access to a connection group, you can update the following attributes:
- Group name
- Identifier Tags (these can never be empty, but they can be completely replaced)
- Connections (a group can be empty)

### Delete a connection group

To delete a connection group:

1. Hover over the group you want to delete and click the **delete (trash can)** icon.
1. Click **Delete**.

> <div class="alert alert-warning"><strong>Note</strong>: Deleting a connection group will impact any workflows that are using the group.</div>

Author: ```suggestion
<div class="alert alert-warning"><strong>Note</strong>: Deleting a connection group impacts any workflows that are using that group.</div>
```

To connect to an arbitrary service, use the HTTP connection type. For authentication options and setup instructions, see [HTTP action][10].

## Connection identifier tags

You can add identifier tags to connections. The tagging rules for connections are based on [Datadog tags][13], with the following additional requirements:

> - Identifier tags must follow the format `tag:value`, and additional colons are not allowed. For example, the identifier tags `env:staging:east` and `env` are invalid formats for connection tags.

Reviewer 3: ```suggestion
- Identifier tags must follow the format `<TAG>:<VALUE>`, and additional colons are not allowed. For example, the identifier tags `env:staging:east` and `env` are invalid formats for connection tags.
```
I actually think you should reject this suggestion, it's technically our style but it obscures things unnecessarily here when there's no potential for confusion anyway IMO. But providing it in case you want to do it by the book!

Author: Acknowledged :)

    - Exactly one colon
- `default` is a reserved value for connection identifier tags. It can't be used as a stand-alone tag key or as a tag value. For example, `default:yes` and `aws:default` are invalid for connection tags.

## Connection groups

You can create groups of connections so that your workflows can authenticate into the correct account or accounts based on the given inputs. Connections can be grouped together only if they share the same integration (for example, you cannot group GCP and AWS connections within the same group). 

You define the members of a connection group using a connection's _Identifier Tags_. For example, you can create a connection group consisting of AWS accounts that have the `account_id` tag.

Each connection in the group must have a set of unique identifier tags so that a workflow can dynamically select the correct connection at runtime. For example:
- `connectionA {account_id:123456789}` and `connectionB {account_id:987654321}` can be grouped together.
- `connectionA {account_id:123456789}` and `connectionC {account_id:123456789}` can't be grouped, because the group would contain duplicate tag values.

### Create a connection group

> <div class="alert alert-info"><strong>Note</strong>:You can only add connections to a group if you have <a href="/service_management/workflows/access/#restrict-access-on-a-specific-connection">Resolver permission</a> for them.</div>

Author: ```suggestion
<div class="alert alert-info"><strong>Note</strong>: You can only add connections to a group if you have <a href="/service_management/workflows/access/#restrict-access-on-a-specific-connection">Resolver permission</a> for them.</div>
```

1. Under **Confirm Group**, use the checkboxes to select the specific members of your group.
1. Click **Next, Confirm Access**, then choose your desired access level for the group.
1. Click **Create**.

### Use a connection group

To use a connection group:

1. In your workflow, select an action that requires a connection.
1. In the **Connection** field, in the drop-down, select the desired connection group under **Groups**.
1. Fill in the desired values for the connection group **Identifiers**. For example, if your connection group is defined using the `env` Identifier Tag, and you have two environments, `prod` and `staging`, you could use either of those values (or an expression that evaluates to one of those values).
1. Fill in any other required step values, then click **Save** to save your workflow.

**Note**: You can only use connections within a group if you have [Resolver permission][12] for those connections. If a workflow tries to use a connection you don't have Resolver permission for, the workflow fails with a `403 Forbidden` error. To fix this issue, you can:
- Configure your workflow so that it can't point to a connection you don't have Resolver permission for.

> - Remove the connection you don't have Resolver permission for from the connection group.<div class="alert alert-danger"><strong>Note</strong>: If you are using a connection group for multiple workflows, removing a connection that another workflow relies on causes that workflow to fail.</div>

Reviewer 3: ```suggestion
- Remove the connection that doesn't have Resolver permission for the connection group.<div class="alert alert-danger"><strong>Note</strong>: If you are using a connection group for multiple workflows, removing a connection that another workflow relies on causes that workflow to fail.</div>
```
Does this still work? I stumbled over "for from" and am trying to revise it to avoid that, but am not sure I captured the intended meaning.

Reviewer 3: Note that in the preview, this alert doesn't render with the correct font size -- I don't think we're expected to include an alert in steps, since usually it goes at the top of a section (even if I like your approach better since the user will have better context where you included it). It's a very minor difference and I don't think it needs to be a stopship concern at all, just flagging in case that matters to you.

Author: Yeah, I noticed the font size thing.

I think the meaning changes if you remove the word, but how about...

> - Remove the connection you don't have Resolver permission for from the connection group.<div class="alert alert-danger"><strong>Note</strong>: If you are using a connection group for multiple workflows, removing a connection that another workflow relies on causes that workflow to fail.</div>

Author: ```suggestion
- Remove the connection for which you don't have Resolver permission from the connection group. <div class="alert alert-danger"><strong>Note</strong>: If you are using a connection group for multiple workflows, removing a connection that another workflow relies on causes that workflow to fail.</div>
```
